### PR TITLE
[SPARK-15911][SQL] Remove the additional Project to be consistent with SQL

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

Currently In `DataFrameWriter`'s `insertInto` and `ResolveRelations` of `Analyzer`, we add additional Project to adjust column ordering. However, it should be using ordering not name for this resolution. This is how Hive does for dynamic partition.
## How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
@cloud-fan There is a test ("Detect table partitioning with correct partition order") in `InsertIntoHiveTableSuite` which is dedicated to test `insertInto` with this column re-ordering. What you think we should do about it? Remove it?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala [hunk 1]
```diff
@@ -257,27 +257,6 @@ class InsertIntoHiveTableSuite extends QueryTest with TestHiveSingleton with Bef
     }
   }
 
-  test("Detect table partitioning with correct partition order") {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala]: can you link to the PR that added this test?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 2]
```diff
@@ -480,10 +480,10 @@ class Analyzer(
                 inputPartCols.find(_.name == name).getOrElse(
                   throw new AnalysisException(s"Cannot find partition column $name"))
               }
+              // Assume partition columns are correctly placed at the end of the child's output
               i.copy(
                 table = EliminateSubqueryAliases(table),
-                partition = tablePartitionNames.map(_ -> None).toMap,
-                child = Project(columns ++ partColumns, child))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Looks like the else branch can be simplified further more?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala (hunk 1 above)
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/InsertIntoHiveTableSuite.scala]: It is added by PR #12239.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: yea. simplified now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 3]
```diff
@@ -470,20 +470,16 @@ class Analyzer(
...
-              // columns to the end of the column list, in partition order.
-              val (inputPartCols, columns) = child.output.partition { attr =>
-                tablePartitionNames.contains(attr.name)
-              }
               // All partition columns are dynamic because this InsertIntoTable had no partitioning
-              val partColumns = tablePartitionNames.map { name =>
-                inputPartCols.find(_.name == name).getOrElse(
-                  throw new AnalysisException(s"Cannot find partition column $name"))
+              tablePartitionNames.filterNot { name =>
+                child.output.exists(_.name == name)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: do we really need this check?

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: hmm. indeed. As we use ordering not name, this check is not needed anymore. I will remove it.

**Comment by viirya** [comment]
@cloud-fan any other thoughts?

**Comment by cloud-fan** [comment]
hi @viirya , we are auditing the insertion behaviour of spark sql, and will have an agreement this week. How about we revisit this PR  after that?

**Comment by viirya** [comment]
@cloud-fan no problem at all.

**Comment by cloud-fan** [comment]
please track the process of https://issues.apache.org/jira/browse/SPARK-16032 and update your PR, thanks!

**Comment by viirya** [comment]
@cloud-fan Looks like a part of this PR is done by some PRs in that JIRA. I will update this.

**Comment by viirya** [comment]
@cloud-fan Updated. Please take a look. Thanks!

**Comment by viirya** [comment]
ping @cloud-fan

**Comment by viirya** [comment]
@cloud-fan Looks like the change in this PR is done by #13766. Let me close it now.

**Pull request closed without merging**